x/oracle/keeper/msg_server.go

```diff
@@ -0,0 +1,203 @@
+package keeper
+
+import (
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+
+	"github.com/persistenceOne/persistence-sdk/x/oracle/types"
+)
+
+type msgServer struct {
+	Keeper
+}
+
+// NewMsgServerImpl returns an implementation of the oracle MsgServer interface
+// for the provided Keeper.
+func NewMsgServerImpl(keeper Keeper) types.MsgServer {
+	return &msgServer{Keeper: keeper}
+}
+
+func (ms msgServer) AggregateExchangeRatePrevote(
+	goCtx context.Context,
+	msg *types.MsgAggregateExchangeRatePrevote,
+) (*types.MsgAggregateExchangeRatePrevoteResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	valAddr, err := sdk.ValAddressFromBech32(msg.Validator)
+	if err != nil {
+		return nil, err
+	}
+
+	feederAddr, err := sdk.AccAddressFromBech32(msg.Feeder)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := ms.ValidateFeeder(ctx, feederAddr, valAddr); err != nil {
+		return nil, err
+	}
+
+	// Ensure prevote wasn't already submitted
+	if ms.HasAggregateExchangeRatePrevote(ctx, valAddr) {
+		return nil, types.ErrExistingPrevote
+	}
+
+	// Convert hex string to votehash
+	voteHash, err := types.AggregateVoteHashFromHexString(msg.Hash)
+	if err != nil {
+		return nil, sdkerrors.Wrap(types.ErrInvalidHash, err.Error())
+	}
+
+	aggregatePrevote := types.NewAggregateExchangeRatePrevote(voteHash, valAddr, uint64(ctx.BlockHeight()))
+	ms.SetAggregateExchangeRatePrevote(ctx, valAddr, aggregatePrevote)
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeAggregatePrevote,
+			sdk.NewAttribute(types.EventAttrKeyVoter, msg.Validator),
+		),
+		sdk.NewEvent(
+			sdk.EventTypeMessage,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.EventAttrValueCategory),
+			sdk.NewAttribute(sdk.AttributeKeySender, msg.Feeder),
+		),
+	})
+
+	return &types.MsgAggregateExchangeRatePrevoteResponse{}, nil
+}
+
+func (ms msgServer) AggregateExchangeRateVote(
+	goCtx context.Context,
+	msg *types.MsgAggregateExchangeRateVote,
+) (*types.MsgAggregateExchangeRateVoteResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	valAddr, err := sdk.ValAddressFromBech32(msg.Validator)
+	if err != nil {
+		return nil, err
+	}
+
+	feederAddr, err := sdk.AccAddressFromBech32(msg.Feeder)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := ms.ValidateFeeder(ctx, feederAddr, valAddr); err != nil {
+		return nil, err
+	}
+
+	params := ms.GetParams(ctx)
+
+	aggregatePrevote, err := ms.GetAggregateExchangeRatePrevote(ctx, valAddr)
+	if err != nil {
+		return nil, sdkerrors.Wrap(types.ErrNoAggregatePrevote, msg.Validator)
+	}
+
+	// Check a msg is submitted proper period
+	if (uint64(ctx.BlockHeight())/params.VotePeriod)-(aggregatePrevote.SubmitBlock/params.VotePeriod) != 1 {
```

Comment:
It can be done as you mentioned but there is one condition.
if `blockHeight = 9` and `submitBlock = 9` then `currentBlock(9) - submitBlock(9) > votePeriod(5)` will also be true.
But as per the current condition, this should error out. So basically current condition only allows a difference of `1 to 5` between `blockHeight and submitBlock`. 

We can have a separate discussion on this.